### ISSUE TITLE
code

### ISSUE THREAD

**Issue by ZhigangSu** [ZhigangSu/Zhigang-Su#5]
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import torch
import torch.nn as nn
from torch.utils.data import Dataset, DataLoader
from sklearn.preprocessing import MinMaxScaler
from sklearn.metrics import mean_squared_error, r2_score

df = pd.read_csv("/mnt/data/psspredict10.csv")
df['date'] = pd.to_datetime(df['date'])
df.sort_values('date', inplace=True)
df.dropna(inplace=True)
df.set_index('date', inplace=True)

data = df['temp'].values.reshape(-1, 1)
scaler = MinMaxScaler(feature_range=(0, 1))
scaled_data = scaler.fit_transform(data)

window_size = 30
def create_dataset(data, window_size):
    X, y = [], []
    for i in range(len(data) - window_size):
        X.append(data[i:i+window_size])
        y.append(data[i+window_size])
    return np.array(X), np.array(y)
X, y = create_dataset(scaled_data, window_size)

train_size = int(len(X) * 0.8)
X_train, y_train = X[:train_size], y[:train_size]
X_test, y_test = X[train_size:], y[train_size:]

X_train_tensor = torch.FloatTensor(X_train)
y_train_tensor = torch.FloatTensor(y_train)
X_test_tensor = torch.FloatTensor(X_test)
y_test_tensor = torch.FloatTensor(y_test)

class TimeSeriesDataset(Dataset):
    def __init__(self, X, y):
        self.X = X
        self.y = y
    def __len__(self):
        return len(self.X)
    def __getitem__(self, idx):
        return self.X[idx], self.y[idx]
batch_size = 32
train_dataset = TimeSeriesDataset(X_train_tensor, y_train_tensor)
train_loader = DataLoader(train_dataset, batch_size=batch_size, shuffle=True)

class LSTMModel(nn.Module):
    def __init__(self, input_size=1, hidden_size=64, num_layers=2, dropout=0.2):
        super(LSTMModel, self).__init__()
        self.lstm = nn.LSTM(input_size, hidden_size, num_layers, batch_first=True, dropout=dropout)
        self.fc = nn.Linear(hidden_size, 1)
    def forward(self, x):
        out, _ = self.lstm(x)
        out = out[:, -1, :]
        out = self.fc(out)
        return out

class BiLSTMModel(nn.Module):
    def __init__(self, input_size=1, hidden_size=64, num_layers=2, dropout=0.2):
        super(BiLSTMModel, self).__init__()
        self.lstm = nn.LSTM(input_size, hidden_size, num_layers, batch_first=True, dropout=dropout, bidirectional=True)
        self.fc = nn.Linear(hidden_size * 2, 1)
    def forward(self, x):
        out, _ = self.lstm(x)
        out = out[:, -1, :]
        out = self.fc(out)
        return out

device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
print("Using device:", device)
model_lstm = LSTMModel().to(device)
model_bilstm = BiLSTMModel().to(device)
criterion = nn.MSELoss()
optimizer_lstm = torch.optim.Adam(model_lstm.parameters(), lr=0.001)
optimizer_bilstm = torch.optim.Adam(model_bilstm.parameters(), lr=0.001)

def train_model(model, optimizer, num_epochs=50):
    model.train()
    for epoch in range(num_epochs):
        total_loss = 0
        for batch_X, batch_y in train_loader:
            batch_X = batch_X.to(device)
            batch_y = batch_y.to(device)
            optimizer.zero_grad()
            outputs = model(batch_X)
            loss = criterion(outputs, batch_y)
            loss.backward()
            optimizer.step()
            total_loss += loss.item() * batch_X.size(0)
        avg_loss = total_loss / len(train_dataset)
        print(f"Epoch {epoch+1}/{num_epochs}, Loss: {avg_loss:.6f}")

print("Training LSTM model...")
train_model(model_lstm, optimizer_lstm, num_epochs=50)
print("Training BiLSTM model...")
train_model(model_bilstm, optimizer_bilstm, num_epochs=50)

def evaluate_model(model, X_test_tensor, y_test_tensor):
    model.eval()
    with torch.no_grad():
        X_test_tensor = X_test_tensor.to(device)
        predictions = model(X_test_tensor)
    predictions = predictions.cpu().numpy()
    y_true = y_test_tensor.numpy()
    predictions_inv = scaler.inverse_transform(predictions)
    y_true_inv = scaler.inverse_transform(y_true)
    mse = mean_squared_error(y_true_inv, predictions_inv)
    r2 = r2_score(y_true_inv, predictions_inv)
    return predictions_inv, y_true_inv, mse, r2

pred_lstm, y_true_inv, mse_lstm, r2_lstm = evaluate_model(model_lstm, X_test_tensor, y_test_tensor)
print("LSTM Test MSE:", mse_lstm, "R2:", r2_lstm)
pred_bilstm, y_true_inv, mse_bilstm, r2_bilstm = evaluate_model(model_bilstm, X_test_tensor, y_test_tensor)
print("BiLSTM Test MSE:", mse_bilstm, "R2:", r2_bilstm)

plt.figure(figsize=(12,6))
plt.plot(y_true_inv, label='True Temperature')
plt.plot(pred_bilstm, label='BiLSTM Predicted Temperature')
plt.title("Test Set Temperature Prediction")
plt.xlabel("Sample")
plt.ylabel("Temperature")
plt.legend()
plt.show()

def predict_future(model, scaled_data, window_size, future_steps):
    model.eval()
    last_window = scaled_data[-window_size:].copy()
    future_preds = []
    for _ in range(future_steps):
        input_tensor = torch.FloatTensor(last_window).unsqueeze(0).to(device)
        with torch.no_grad():
            pred = model(input_tensor)
        pred_val = pred.cpu().numpy()[0, 0]
        future_preds.append(pred_val)
        last_window = np.append(last_window[1:], [[pred_val]], axis=0)
    future_preds = np.array(future_preds).reshape(-1, 1)
    future_preds_inv = scaler.inverse_transform(future_preds)
    return future_preds_inv

future_steps = 180
future_predictions = predict_future(model_bilstm, scaled_data, window_size, future_steps)
plt.figure(figsize=(12,6))
plt.plot(future_predictions, label='Predicted Temperature for Future 6 Months')
plt.title("Future 6 Months Temperature Prediction")
plt.xlabel("Day")
plt.ylabel("Temperature")
plt.legend()
plt.show()
